src/Workspaces/Core/Portable/Storage/SQLite/SQLitePersistentStorage.Accessor.cs

```diff
@@ -136,22 +160,55 @@ private Task FlushPendingWritesAsync(TKey key, CancellationToken cancellationTok
             private Task AddWriteTaskAsync(TKey key, Action<SqlConnection> action, CancellationToken cancellationToken)
                 => Storage.AddWriteTaskAsync(_writeQueueKeyToWrites, GetWriteQueueKey(key), action, cancellationToken);
 
-            private Stream ReadBlob(SqlConnection connection, TDatabaseId dataId)
+            private Stream ReadBlob(
+                SqlConnection connection, TDatabaseId dataId, string columnName,
+                Checksum checksumOpt, CancellationToken cancellationToken)
             {
+                // Note: it's possible that someone may write to this row between when we
+                // get the row ID above and now.  That's fine.  We'll just read the new
+                // bytes that have been written to this location.  Note that only the
+                // data for a row in our system can change, the ID will always stay the
+                // same, and the data will always be valid for our ID.  So there is no
+                // safety issue here.
                 if (TryGetRowId(connection, dataId, out var rowId))
                 {
-                    // Note: it's possible that someone may write to this row between when we
-                    // get the row ID above and now.  That's fine.  We'll just read the new
-                    // bytes that have been written to this location.  Note that only the
-                    // data for a row in our system can change, the ID will always stay the
-                    // same, and the data will always be valid for our ID.  So there is no
-                    // safety issue here.
-                    return connection.ReadBlob(DataTableName, DataColumnName, rowId);
+                    // Have to run the blob reading in a transaction.  This is necessary
+                    // for two reasons.  First, blob reading outside a transaction is not
+                    // safe to do with the sqlite API.  It may produce corrupt bits if 
+                    // another thread is writing to the blob.  Second, if a checksum was
+                    // passed in, we need to validate that the checksums match.  This is
+                    // only safe if we are in a transaction and no-one else can race with
+                    // us.
+                    Stream stream = null;
+                    connection.RunInTransaction(() =>
+                    {
+                        // If we were passed a checksum, make sure it matches what we have
+                        // stored in the table already.  If they don't match, don't read
+                        // out the data value at all.
+                        if (checksumOpt != null &&
```

Comment:
I see, this is why checksum must be passed in since we don't support transaction from outside. in that case, shouldn't WriteAsync do what readAsync does and remove GetChecksumAsync? since thing can change as checksum is checked and then when it is written?

Reply:
I guess we can ignore such case since one of them win and overwrite data anyway.

Reply:
>  in that case, shouldn't WriteAsync do what readAsync does and remove GetChecksumAsync? since thing can change as checksum is checked and then when it is written?

I don't understand the question about WriteAsync.  There is no need to check anything.  When you write, you're saying "write out this entire row".  That has no need for checks.  It just happens atomically.  When you read, you actually need to be in a transaction so you can validate the checksum in a consistent fashion and not have race conditions happen.

Reply:
okay, why ReadChecksumAsync is provided if it can cause a race?